File: js-version/js/locate.js
(function (root) {
  const isNode = typeof module === "object" && module.exports;
  const LRS = isNode ? require("./schema.js") : root.LRS || (root.LRS = {});

  function locatePoints(points, events, options = {}) {
    const how = options.how || "inner";
    if (!["inner", "left"].includes(how)) throw new Error("how must be 'inner' or 'left'");

    const pts = LRS.cloneRows(points);
    let ev = LRS.cloneRows(events);
    const pointSchema = options.pointSchema || LRS.LrsSchema.fromRows(pts, { requireMeasure: true });
    const eventSchema = options.eventSchema || LRS.LrsSchema.fromRows(ev, { requireLine: true });
    pointSchema.validatePoint(pts);
    eventSchema.validateLine(ev);

    pts.forEach((row) => {
      row._merge_roadway = LRS.padRoadwayId(row[pointSchema.roadway]);
      row[pointSchema.measure] = LRS.coerceNumeric(row[pointSchema.measure]);
    });
    ev.forEach((row) => {
      row._merge_roadway = LRS.padRoadwayId(row[eventSchema.roadway]);
      row[eventSchema.bmp] = LRS.coerceNumeric(row[eventSchema.bmp]);
      row[eventSchema.emp] = LRS.coerceNumeric(row[eventSchema.emp]);
    });

    const evCols = LRS.columnsOf(ev);
    let keepEvent;
    if (options.eventCols == null) {
      keepEvent = evCols.filter((col) => col !== eventSchema.roadway && col !== "_merge_roadway" && col !== "geometry");
    } else {
      keepEvent = options.eventCols.filter((col) => evCols.includes(col) && col !== "geometry");
      for (const required of [eventSchema.bmp, eventSchema.emp]) {
        if (!keepEvent.includes(required) && evCols.includes(required)) keepEvent.push(required);
      }
    }

    const protectPoint = new Set([
      pointSchema.roadway,
      pointSchema.measure,
      "_merge_roadway",
      "LRS_PARENT_ID",
      "LRS_SUB_ID",
      "LRS_SOURCE_ROW",
      "ROADWAY_ORIG",
    ]);

    function coversMeasure(eventRow, measure, onRoad) {
      const bmp = eventRow[eventSchema.bmp];
      const emp = eventRow[eventSchema.emp];
      if (measure == null || bmp == null || emp == null) return false;
      if (measure < bmp || measure > emp) return false;
      if (measure === emp && measure > bmp) {
        const startsNext = onRoad.some((other) => other !== eventRow && other[eventSchema.bmp] === measure);
        if (startsNext) return false;
      }
      return true;
    }

    const joined = [];
    for (const point of pts) {
      const onRoad = ev.filter((row) => {
        if (row._merge_roadway !== point._merge_roadway) return false;
        for (const pair of options.matchPairs || []) {
          if (!pair || !pair.point || !pair.event) continue;
          if (LRS.compareKey(point[pair.point]) !== LRS.compareKey(row[pair.event])) return false;
        }
        return true;
      });
      const matches = onRoad.filter((row) => coversMeasure(row, point[pointSchema.measure], onRoad));
      const partners = matches.length ? matches : [null];
      for (const eventRow of partners) {
        const out = LRS.dropColumns(point, ["_merge_roadway"]);
        for (const col of keepEvent) {
          if (protectPoint.has(col) || col === "geometry") continue;
          out[col] = eventRow ? eventRow[col] : null;
        }
        const fromPoint = LRS.isPointGeom?.(point.geometry) ? point.geometry : null;
        const fromLine =
          eventRow && LRS.pointAtMeasure
            ? LRS.pointAtMeasure(
                eventRow.geometry,
                eventRow[eventSchema.bmp],
                eventRow[eventSchema.emp],
                point[pointSchema.measure]
              )
            : null;
        if (fromPoint || fromLine) out.geometry = fromPoint || fromLine;
        else delete out.geometry;
        joined.push(out);
      }
    }

    const located = [];
    const unmatched = [];
    for (const row of joined) {
      const measure = LRS.coerceNumeric(row[pointSchema.measure]);
      const bmp = LRS.coerceNumeric(row[eventSchema.bmp]);
      const emp = LRS.coerceNumeric(row[eventSchema.emp]);
      const onSegment = measure != null && bmp != null && emp != null && measure >= bmp && measure <= emp;
      if (onSegment) located.push(row);
      else unmatched.push(row);
    }
    return { located, unmatched };
  }

  Object.assign(LRS, { locatePoints });
  if (isNode) module.exports = LRS;
})(typeof globalThis !== "undefined" ? globalThis : this);

File: python-version/lrs_tools/locate.py
"""Locate point events onto line events using roadway ID + milepost."""

from __future__ import annotations

import pandas as pd

from lrs_tools.schema import LrsSchema, coerce_numeric, pad_roadway_id


def locate_points(
    points: pd.DataFrame,
    events: pd.DataFrame,
    point_schema: LrsSchema | None = None,
    event_schema: LrsSchema | None = None,
    *,
    how: str = "inner",
    event_cols: list[str] | None = None,
) -> tuple[pd.DataFrame, pd.DataFrame]:
    """Join points to line events when ``BMP <= LOCATION <= EMP`` on the same roadway.

    Returns ``(located, unmatched)``. *unmatched* is the roadway-join explosion
    that failed the milepost test, matching the existing Step 2 QC export.
    """
    if how not in {"inner", "left"}:
        raise ValueError("how must be 'inner' or 'left'")

    point_schema = point_schema or LrsSchema.from_dataframe(points, require_measure=True)
    event_schema = event_schema or LrsSchema.from_dataframe(events, require_line=True)
    point_schema.validate_point(points)
    event_schema.validate_line(events)

    pts = points.copy()
    ev = events.copy()
    if "geometry" in ev.columns:
        ev = ev.drop(columns=["geometry"])

    pts["_merge_roadway"] = pts[point_schema.roadway].map(pad_roadway_id)
    ev["_merge_roadway"] = ev[event_schema.roadway].map(pad_roadway_id)
    pts[point_schema.measure] = coerce_numeric(pts[point_schema.measure])
    ev[event_schema.bmp] = coerce_numeric(ev[event_schema.bmp])
    ev[event_schema.emp] = coerce_numeric(ev[event_schema.emp])

    if event_cols is None:
        keep_event = [
            col
            for col in ev.columns
            if col not in {event_schema.roadway, "_merge_roadway"}
        ]
    else:
        keep_event = [col for col in event_cols if col in ev.columns]
        for required in (event_schema.bmp, event_schema.emp):
            if required not in keep_event and required in ev.columns:
                keep_event.append(required)

    ev_slim = ev[["_merge_roadway"] + keep_event].copy()
    joined = pts.merge(ev_slim, on="_merge_roadway", how="left")

    on_segment = (
        joined[point_schema.measure].notna()
        & joined[event_schema.bmp].notna()
        & joined[event_schema.emp].notna()
        & (joined[point_schema.measure] >= joined[event_schema.bmp])
        & (joined[point_schema.measure] <= joined[event_schema.emp])
    )
    located = joined.loc[on_segment].drop(columns=["_merge_roadway"]).reset_index(drop=True)
    unmatched = joined.loc[~on_segment].drop(columns=["_merge_roadway"]).reset_index(drop=True)

    if how == "left" and located.empty and unmatched.empty:
        return joined.drop(columns=["_merge_roadway"]), unmatched
    return located, unmatched
